File: Week_09/5/longest_palindromic_substring.go
package main

func longestPalindrome(s string) string {
	ans := ""

	dp := make([][]bool, len(s))
	for i := 0; i < len(s); i++ {
		dp[i] = make([]bool, len(s))
	}

	for l := 0; l < len(s); l++ {
		for i := 0; i+l < len(s); i++ {
			j := i + l
			if l == 0 {
				dp[i][j] = true
			} else if l == 1 {
				dp[i][j] = s[i] == s[j]
			} else {
				dp[i][j] = s[i] == s[j] && dp[i+1][j-1]
			}

			if dp[i][j] && l+1 > len(ans) {
				ans = s[i : i+l+1]
			}
		}
	}
	return ans
}

File: Week_09/5/longest_palindromic_substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        dp = [[False] * len(s) for _ in range(len(s))]

        ans = ""
        for l in range(0, len(s)):
            for i in range(0, len(s) - l):
                j = i + l
                if l == 0:
                    dp[i][j] = True

                elif l == 1:
                    dp[i][j] = s[i] == s[j]
                else:
                    dp[i][j] = s[i] == s[j] and dp[i + 1][j - 1]

                if dp[i][j] and l + 1 > len(ans):
                    ans = s[i:j + 1]
        return ans